# **Getting started with bioinformatics coding** <br />



<br />
<p align="center">
  <img 
    src="https://github.com/asadprodhan/Getting-started-with-bioinformatics-coding/blob/main/TerminalScreen2.PNG"
  >
</p>
<p align = "center">
A simple bioinformatics script. It filters out the incomplete or corrupted fastq reads.
</p>



**Q: I want to do bioinformatics coding. Which programming language do I need to learn?**


It depends on the task. However, three programming languages- R, Python and Bash- can provide with all the command line tools required for High Throughput Sequencing (HTS) based bioinformatics projects. For example, disease diagnosis and surveillance, species identification through metagenomic analysis of the eDNA (Environmental DNA), phylogenetic analysis, RNA-seq analysis, Genotyping-by-sequencing (GBS) data analysis, SNP marker development, linkage map construction, QTL mapping etc. 


***R*** is one of the best programming languages for data visualisation.


***Python*** is famous for its data wrangling ability.


***Bash*** is an interpreter shell. It means that it takes your commands that you type in the terminal and calls the Operating System (OS) to execute those commands. For example, when you type ‘ls’ in the terminal, the interpreter takes it and ask the Linux OS to list the content of the current directory. However, Bash is also an essential programming language. In fact, it is one of the ‘must’ skills if you want to get a good handle on the bioinformatics coding.


**Q: Where can I learn these languages?** 


There are abundant resources freely available online. 


A good starting point is to do a Google and YouTube search. Here are some examples: 


**R**


- R Graphics Cookbook: Practical Recipes for Visualizing Data By Winston Chang 
- HarvardX Data Science Training (http://rafalab.github.io/pages/harvardx.html)
- Bioconductor (https://kasperdanielhansen.github.io/genbioconductor/) 


**Python**


- YouTube (Python Tutorials for Absolute Beginners by CS Dojo)
- Computing for Biologists (https://people.duke.edu/~ccc14/pcfb/python1.html)
- TutorialsTeacher (https://www.tutorialsteacher.com/python)
- SoloLearn (Mobile app to learn Python)
- Biopython (http://biopython.org/DIST/docs/tutorial/Tutorial.html)
- Coursera (https://www.coursera.org/learn/python-genomics) 


**Bash**


- YouTube (OMGenomics, https://www.youtube.com/watch?v=EMaFdfIlK58)
- Software carpentry (https://software-carpentry.org/lessons/)


**Q: How do I know what bioinformatic tools I need for my HTS data analyses?**


- Do a literature search in Google and Google Scholar using Keywords
- Find a few publications that match with your project objectives
- List down what bioinformatic tools have been used in those publications
- Then, compare the requirements and dependencies of those tools to find out the best suitable one matching your programming skills and interest


**Q: Where can I get the above selected bioinformatics tools?**


- Anaconda
- R repository like CRAN, and 
- biocontainers (quay.io) are the popular distributors for data science including bioinformatics tools/packages


**Q: It feels overwhelming. What is the effective learning strategy in bioinformatics?**


It is common that learning bioinformatics feels overwhelming. This is partly because there are so many resources and tools to accomplish the same task. Do I have to learn all of them? ‘NO’. Pick up one and learn it like you will be teaching it in the end. 



***My learning philosophy is ‘first, get a clear goal, then find and learn the relevant tool to accomplish it’. If I go the other way around, I feel overwhelming and cannot accomplish my project goals efficiently.***


